cleverhans/attacks.py

```diff
@@ -1,4 +1,374 @@
+        :param clip_min: (optional float) Minimum input component value
+        :param clip_max: (optional float) Maximum input component value
+        """
+        super(BasicIterativeMethod, self).__init__(model, back, sess, params)
+
+        # Check that all required attack specific parameters are defined
+        req = ('eps', 'eps_iter', 'nb_iter', 'y')
+        if not all(k in params for k in req):
+            raise Exception("Attack requires label placeholder.")
+
+        # Save attack-specific parameters
+        self.eps = params['eps']
+        self.eps_iter = params['eps_iter']
+        self.nb_iter = params['nb_iter']
+        self.y = params['y']
+        self.ord = params['ord'] if 'ord' in params else np.inf
+        self.clip_min = params['clip_min'] if 'clip_min' in params else None
+        self.clip_max = params['clip_max'] if 'clip_max' in params else None
+
+        # Check if order of the norm is acceptable given current implementation
+        if self.ord not in [np.inf, 1, 2]:
+            raise Exception("Norm order must be either np.inf, 1, or 2.")
+        if back == 'th':
+            error_string = "BasicIterativeMethod is not implemented in Theano"
+            raise NotImplementedError(error_string)
+
+    def generate(self, x, params={}):
+        import tensorflow as tf
+
+        # Initialize loop variables
+        eta = 0
+
+        # Fix labels to the first model predictions for loss computation
+        model_preds = self.model(x)
+        preds_max = tf.reduce_max(model_preds, 1, keep_dims=True)
+        y = tf.to_float(tf.equal(model_preds, preds_max))
+
+        for i in range(self.nb_iter):
+            FGSM = FastGradientMethod(self.model, back=self.back,
+                                      sess=self.sess,
+                                      params={'eps': self.eps_iter,
+                                              'y': y})
+            # Compute this step's perturbation
+            eta = FGSM.generate(x + eta) - x
+
+            # Clipping perturbation eta to self.ord norm ball
+            if self.ord == np.inf:
+                eta = tf.clip_by_value(eta, -self.eps, self.eps)
+            elif self.ord in [1, 2]:
+                reduc_ind = list(xrange(1, len(eta.get_shape())))
+                if self.ord == 1:
+                    norm = tf.reduce_sum(tf.abs(eta),
+                                         reduction_indices=reduc_ind,
+                                         keep_dims=True)
+                elif self.ord == 2:
+                    norm = tf.sqrt(tf.reduce_sum(tf.square(eta),
+                                                 reduction_indices=reduc_ind,
+                                                 keep_dims=True))
+                eta = eta * self.eps / norm
+
+        # Define adversarial example (and clip if necessary)
+        adv_x = x + eta
+        if self.clip_min is not None and self.clip_max is not None:
```

Comment:
The loop is iterating over the eta perturbation, so if we move this back in the loop, we need to compute eta back after the clipping is done